dds/DCPS/DataReaderImpl.cpp

```diff
@@ -3245,6 +3245,18 @@ DataReaderImpl::unregister_for_writer(const RepoId& participant,
   TransportClient::unregister_for_writer(participant, readerid, writerid);
 }
 
+void
+DataReaderImpl::update_locators(const RepoId& writerId,
+                                const TransportLocatorSeq& locators)
+{
+  ACE_READ_GUARD(ACE_RW_Thread_Mutex, read_guard, writers_lock_);
+  WriterMapType::const_iterator iter = writers_.find(writerId);
+
+  if (iter != writers_.end()) {
+    TransportClient::update_locators(writerId, locators);
```

Comment:
I got rid of the global lock in DataWriterImpl.  By holding the locks, we ensure that the transport stays in sync with the DataReader/DataWriter.  Otherwise, we can leak locators in the transport.

As to the locking...well...you know.